File: src/gens/swift/TcpStream.swift
#if os(Linux)
	import Glibc
#else
	import Darwin
#endif

class TcpStream : InputStream, OutputStream {
    init(_ host: String, _ port: Int) {
        sock = socket(AF_INET, Int32(SOCK_STREAM.rawValue), 0)
        if sock == -1 {
            fatalError("Failed to create socket")
        }
        var yes: Int32 = 1
        if setsockopt(sock, Int32(IPPROTO_TCP), TCP_NODELAY, &yes, socklen_t(MemoryLayout<Int32>.size)) < 0 {
            fatalError("Failed to set TCP_NODELAY")
        }
        var hints = addrinfo()
        var servinfo: UnsafeMutablePointer<addrinfo>?
        hints.ai_family = AF_INET;
        hints.ai_socktype = Int32(SOCK_STREAM.rawValue);
        if (getaddrinfo(host, String(port), &hints, &servinfo) != 0) {
            fatalError("Failed to get addr info")
        }
        if (connect(sock, servinfo!.pointee.ai_addr, servinfo!.pointee.ai_addrlen) == -1) {
            fatalError("Failed to connect")
        }
        freeaddrinfo(servinfo)
    }

	func readBytes(_ byteCount: Int) -> [Byte] {
        var byteCount = byteCount
		var buffer = [Byte](repeating: 0x0, count: byteCount)
        var pos = 0
        while byteCount > 0 {
            let received = buffer[pos...].withUnsafeMutableBytes { recv(sock, $0.baseAddress, byteCount, 0) }
            if received < 0 {
                fatalError("Failed to read from socket")
            }
            pos += received
            byteCount -= received
        }
		return buffer
    }

	func writeBytes(_ data: [Byte]) {
        if send(sock, data, data.count, 0) < 0 {
            fatalError("Failed to write to socket")
        }
    }

    func flush() {
    }

    private let sock: Int32
}

File: generated-code/php/codecraft/TcpReadWrite/TcpStream.php
<?php

require_once 'Stream.php';
require_once 'BufferedStream.php';

class TcpStream
{
    private $socket;
    public $inputStream;
    public $outputStream;
    function __construct($host, $port)
    {
        $this->socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
        if (!socket_set_option($this->socket, SOL_TCP, TCP_NODELAY, 1)) {
            throw new Exception("Failed to set TCP_NODELAY");
        }
        if (!socket_connect($this->socket, $host, $port)) {
            throw new Exception("Failed to connect");
        }
        $this->inputStream = new BufferedInputStream(new TcpInputStream($this->socket));
        $this->outputStream = new BufferedOutputStream(new TcpOutputStream($this->socket));
    }
    function __destruct()
    {
        socket_close($this->socket);
    }
}

class TcpInputStream extends InputStream
{
    private $socket;
    function __construct($socket)
    {
        $this->socket = $socket;
    }
    public function readAtMost($byteCount)
    {
        return socket_read($this->socket, $byteCount);
    }
}

class TcpOutputStream extends OutputStream
{
    private $socket;
    function __construct($socket)
    {
        $this->socket = $socket;
    }
    public function write($bytes)
    {
        socket_write($this->socket, $bytes);
    }
    public function flush()
    {
    }
}
